                                                 Домашнє завдання N 7

Так як ми працюємо з рекомендованними системами, по завданню нам радять використовувати бібліотеку Surprise, котра якраз є по суті додатком до знайомої нам бібліотеки scikit-learn для тренування моделей рекомендаційних систем.

Тому давайте імпортуємо цю бібліотеку і також завантажимо потрібний датасет movielens на 100 к даних. Думаю, нам буде достатньо цих даних для виконання домашнього завдання:

In [4]:
from surprise import Dataset, Reader, SVD, SVDpp, NMF
from surprise.model_selection import cross_validate

In [5]:
dataset_path = './ml-100k/u.data'
reader = Reader(line_format='user item rating timestamp', sep='\t')
data = Dataset.load_from_file(dataset_path, reader=reader)

data

Давайте спробуємо вивести перші 5 рядків, щоб побачити, які дані містить наш датасет. Для цього використаємо трохи інший ніж зазвичай метод виводу:

In [6]:
N = 5
data_raw = data.raw_ratings

for i in range(N):
    print(data_raw[i])



('196', '242', 3.0, '881250949')
('186', '302', 3.0, '891717742')
('22', '377', 1.0, '878887116')
('244', '51', 2.0, '880606923')
('166', '346', 1.0, '886397596')


Окей, супер!

Тепер давайте побудуємо модел за допомогою 3 алгоритмів розграхунків: SVD, SVD++, NMF.

Почнемо з SVD:

In [7]:
algo = SVD()

cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9360  0.9307  0.9292  0.9380  0.9384  0.9345  0.0038  
MAE (testset)     0.7374  0.7366  0.7319  0.7406  0.7400  0.7373  0.0031  
Fit time          0.39    0.40    0.38    0.39    0.40    0.39    0.01    
Test time         0.04    0.04    0.06    0.06    0.04    0.05    0.01    


{'test_rmse': array([0.93596623, 0.93071379, 0.92916717, 0.9380197 , 0.93841958]),
 'test_mae': array([0.73743657, 0.73664825, 0.7319266 , 0.74056852, 0.73999327]),
 'fit_time': (0.39029908180236816,
  0.4037020206451416,
  0.3790609836578369,
  0.3901810646057129,
  0.3962550163269043),
 'test_time': (0.038311004638671875,
  0.03725600242614746,
  0.0580441951751709,
  0.06064009666442871,
  0.03810715675354004)}

Тепер те саме зробимо з SVD++:

In [8]:
algo_svdpp = SVDpp()

cross_validate(algo_svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9200  0.9208  0.9175  0.9180  0.9205  0.9194  0.0013  
MAE (testset)     0.7197  0.7225  0.7192  0.7210  0.7241  0.7213  0.0018  
Fit time          6.09    6.13    6.18    6.15    6.22    6.15    0.04    
Test time         1.25    1.26    1.27    1.28    1.28    1.27    0.01    


{'test_rmse': array([0.91999709, 0.92078641, 0.91752432, 0.91800438, 0.92047225]),
 'test_mae': array([0.71967915, 0.72254102, 0.71920926, 0.72097416, 0.72414852]),
 'fit_time': (6.088117837905884,
  6.129518270492554,
  6.179074048995972,
  6.150877952575684,
  6.2154762744903564),
 'test_time': (1.2491528987884521,
  1.2640938758850098,
  1.2745099067687988,
  1.2801599502563477,
  1.2845590114593506)}

І також використовуючи NMF модель:

In [9]:
algo_nmf = NMF()

cross_validate(algo_nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9671  0.9687  0.9629  0.9582  0.9672  0.9648  0.0038  
MAE (testset)     0.7620  0.7588  0.7551  0.7533  0.7568  0.7572  0.0030  
Fit time          0.50    0.53    0.49    0.48    0.48    0.49    0.02    
Test time         0.06    0.06    0.06    0.03    0.06    0.06    0.01    


{'test_rmse': array([0.96706594, 0.96874293, 0.96292619, 0.95823171, 0.96720778]),
 'test_mae': array([0.76195328, 0.75875549, 0.75505491, 0.75328049, 0.75682189]),
 'fit_time': (0.4980957508087158,
  0.5279371738433838,
  0.48900413513183594,
  0.47606587409973145,
  0.4817488193511963),
 'test_time': (0.06377315521240234,
  0.05922102928161621,
  0.06195688247680664,
  0.030456066131591797,
  0.05999398231506348)}

Ми отримали результати оцінок RMSE та MAE всіх 3 алгоритмів. Давайте проаналізуємо отримані дані та зробимо висновки:

Алгоритм SVD: Має середнє значення RMSE приблизно 0.9345 і MAE приблизно 0.7373. Цей алгоритм показав непогані результати з помірним часом навчання та тестування.

Алгоритм SVD++: Показав кращі результати за RMSE (середнє 0.9194) та MAE (середнє 0.7213) порівняно з SVD, що свідчить про вищу точність прогнозування. Однак, час навчання та тестування значно вищий, що є важливим аспектом для великих наборів даних або обмежених обчислювальних ресурсів.

Алгоритм NMF: Має найвищі значення помилок RMSE (середнє 0.9648) та MAE (середнє 0.7572) серед трьох розглянутих алгоритмів, що вказує на нижчу точність прогнозування порівняно з SVD та SVD++. Час навчання та тестування подібний до SVD.

Як висновок, можемо зазначити, що найбільш вдалим тут здається вибір алгоритму SVD++, так як в нас датасет не такий великий. Якби ми взяли датасет не на 100к а на 1М даних, то досить велика ймовірність, що краще б було тоді застосовувати алгоритм SVD.